mono/metadata/threads.c

```diff
@@ -888,19 +904,19 @@ mono_thread_detach_internal (MonoInternalThread *thread)
 	thread->current_appcontext = NULL;
 
 	/*
-	 * thread->synch_cs can be NULL if this was called after
-	 * ves_icall_System_Threading_InternalThread_Thread_free_internal.
-	 * This can happen only during shutdown.
-	 * The shutting_down flag is not always set, so we can't assert on it.
+	 * This should be alive until after the reference queue runs the
+	 * post-free cleanup function
 	 */
-	if (thread->synch_cs)
-		LOCK_THREAD (thread);
+	while (TRUE) {
```

Comment:
Are there paths that take the lock to cover a state change, and more, and read it multiple times under lock and expect it not to change, and this would introduce  race conditions with those? I'd have to survey the code..

Reply:
I would rather not depend on that. This seems conservative, and remains fast in the correct case.

Reply:
This is not as conservative as it may seem, as it may create race conditions.  Let me check the code..

Reply:
For example,  `mono_thread_clear_and_set_state` and `mono_thread_test_and_set_state` do read/modify/write with lock. How does that interoperate with interlocked changes? You cannot easily/generally have multiple locking protocols for the same data.

Reply:
`mono_thread_suspend` => `async_suspend_internal` similarly.

Reply:
Worth noting is that before this PR was landed, the control flow would have been to change this enum without locking at all. This would have run after the finalizer, which would have already freed it.

Reply:
I think that we should do the opposite, and create helper libraries for our fancy locking patterns that can be proven correct in isolation.

Reply:
And ban use of mutex primitives in most code

Reply:
I think we've been doing just that, slowly